packages/runtime/src/utils/serializeState.ts

```diff
@@ -0,0 +1,30 @@
+import { mapValues } from '@makeswift/controls'
+
+export type SerializedState<T> =
+  T extends Map<string, infer U>
+    ? Record<string, SerializedState<U>>
+    : T extends Record<string, infer U>
+      ? Record<string, SerializedState<U>>
+      : T extends Set<infer U>
+        ? Array<SerializedState<U>>
+        : T extends Array<infer U>
+          ? Array<SerializedState<U>>
+          : T
+
+export function serializeState<T>(state: T): SerializedState<T> {
+  if (state instanceof Map) {
+    return Object.fromEntries(
+      [...state.entries()].map(([key, v]) => [key, serializeState(v)]),
+    ) as SerializedState<T>
+  }
+
+  if (state instanceof Set) return [...state.values()].map(serializeState) as SerializedState<T>
+
+  if (state instanceof Array) return state.map(serializeState) as SerializedState<T>
+
+  if (state !== null && typeof state === 'object') {
+    return mapValues(state, serializeState) as SerializedState<T>
+  }
+
+  return state as SerializedState<T>
+}
```

Comment:
Recursively translate `Map` to `Record`, `Set` to `Array`.

Reply:
Should this utility be named `serializeState` even though it isn't particularly tied to our state? Seems like this is a pretty generic deep `Map` to array of entries and `Set` to array utility

Reply:
It is generic, but not generic enough to be just called `serialize`, plus the details of what it means to serialize something in general often depend on the use case. Our use case is serializing state, thus the name. We could name it something like `serializeMapSet`, but then again, if we decide to add, say `Date` serialization, we'd have to adjust the name.

> Seems like this is a pretty generic deep `Map` to array of entries and `Set` to array utility

`Map` to `Record`, `Set` to `Array`.